src/autograd/Functions.cpp

```diff
@@ -117,6 +117,35 @@ namespace af {
             return Variable(result, false);
         }
 
+        Variable select_index(const Variable &input, const Variable &idx)
+        {
+            af::array result = input.array()(idx.array());
+            af::array mask   = af::constant(0, input.dims());
+            mask(idx.array()) = 1;
+
+            auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
+                auto grad = inputs[2].array();
+                auto grad_mask = af::where(grad);
```

Comment:
Why are you converting `idx` into a mask and converting it back into an index again? This is making unnecessary copies.